packages/hdwallet-keepkey/src/ethereum.ts

```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
 }
 
 export async function ethSignMessage(transport: Transport, msg: core.ETHSignMessage): Promise<core.ETHSignedMessage> {
-  const m = new Messages.EthereumSignMessage();
+  const m = new Ethereum.EthereumSignMessage();
   m.setAddressNList(msg.addressNList);
   m.setMessage(ethers.utils.isBytes(msg.message) ? ethers.utils.arrayify(msg.message) : toUTF8Array(msg.message));
   const response = await transport.call(Messages.MessageType.MESSAGETYPE_ETHEREUMSIGNMESSAGE, m, {
     msgTimeout: core.LONG_TIMEOUT,
   });
-  const sig = response.proto as Messages.EthereumMessageSignature;
+  const sig = response.proto as Ethereum.EthereumMessageSignature;
   return {
     address: eip55.encode("0x" + core.toHexString(sig.getAddress_asU8())), // FIXME: this should be done in the firmware
     signature: "0x" + core.toHexString(sig.getSignature_asU8()),
   };
 }
 
+export async function ethSignTypedData(
+  transport: Transport,
+  msg: core.ETHSignTypedData
+): Promise<core.ETHSignedTypedData> {
+  /**
+   * If the message to be signed is sufficiently small, the KeepKey can calculate the
+   * domain separator and message hashes. Otherwise, we need to pre-calculate hashes
+   * here and verify on device.
+   */
+
+  let result: Ethereum.EthereumTypedDataSignature;
+
+  const sTypes = JSON.stringify({ types: msg.types });
+  const sPrimaryType = JSON.stringify({ primaryType: msg.primaryType });
+  const sDomain = JSON.stringify({ domain: msg.domain });
+  const sMessage = JSON.stringify({ message: msg.message });
+
+  try {
+    if (sTypes.length > 2048 || sPrimaryType.length > 80 || sDomain.length > 2048 || sMessage.length > 2048) {
+      /* Pre-calculate domain separator and messages hashes and verify on KeepKey */
+      const hashableMessage: TypedData = {
```

Comment:
thoughts about nesting this object in `coreETHSignTypedData` to prevent the need to recreate the `TypedData` object?
```
export type ETHSignTypedData = { data: TypedData; addressNList: BIP32Path };
```

Reply:
sTypes, sPrimaryType, sDomain, sMessage are representations of the data in the ETHSignedTypedData object formatted specifically for KeepKey. Personally, I like to keep application-specific logic as close to the application as possible. Here, I think it's cleaner to transform the data for KeepKey in the hdwallet-keepkey adapter instead of extending the shared data type to include fields that are only consumed in one place.

Reply:
Just to clarify, I was meaning this comment to be in reference to `hashableMessage` vs the `sVariables`. 
![image](https://user-images.githubusercontent.com/35275952/201800421-f6299370-3424-4ad7-9d85-5644133ef58a.png)

Reply:
We aren't yet using this type anywhere else, and there isn't any reason why the object necessarily needs to be flat, so I don't see any issue with `core.ETHSignTypedData` having separate fields for the hashable data and addressNList. I'll make that change now.

Reply:
Feels more ergonomic this way to me. Cool, thanks!